certdb/ocspstapling.go

```diff
@@ -0,0 +1,158 @@
+package certdb
+
+import (
+	"bytes"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/binary"
+	"github.com/google/certificate-transparency/go"
+	"golang.org/x/crypto/ocsp"
+)
+
+// OID of the SCT extension
+var sctExtOid = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2}
+
+// TODO: (?) Write function to retrieve a []ct.SignedCertificateTimestamp. How
+// to decide which SCT to return in order to "dynamically select an appropriate
+// single SCT to conserve bandwidth"?
+// https://www.certificate-transparency.org/faq
+
+// StapleSCTList inserts a list of Signed Certificate Timestamps into all OCSP
+// responses in a database wrapped by a given certdb.Accessor.
+//
+// TODO: Add an error return value.
+//
+// NOTE: Returns a bool indicating whether the SCT was successfully inserted
+// into ALL corresponding OCSPRecords (i.e. one failure results in a return
+// value of false).
+// NOTE: This function is patterned after the exported Sign method in
+// https://github.com/cloudflare/cfssl/blob/master/signer/local/local.go
+func StapleSCTList(acc Accessor, serial, aki string, scts []ct.SignedCertificateTimestamp) bool {
+	// Grab all OCSP records that match serial and aki
+	ocspRecs, err := acc.GetOCSP(serial, aki)
+	if err != nil || len(ocspRecs) == 0 {
+		// { There was an error or the OCSPRecord does not exist }
+		return false
+	}
+
+	// Add the SCTs to each ocsp response
+	for _, rec := range ocspRecs {
+		// NOTE: Assuming that rec.Body is in base64-encoded DER form
+		// based on
+		// https://github.com/cloudflare/cfssl/blob/master/ocsp/responder.go#L54
+		der, err := base64.StdEncoding.DecodeString(rec.Body)
+		if err != nil {
+			// { decoding error }
+			return false
+		}
+
+		response, err := ocsp.ParseResponse(der, nil)
+		if err != nil {
+			// { parsing error }
+			return false
+		}
+
+		// Find the SCTListExtension in the ocsp response
+		var SCTListExtension, ext pkix.Extension
+		var idxExt int
+		for idxExt, ext = range response.Extensions {
+			if ext.Id.Equal(sctExtOid) {
+				SCTListExtension = ext
+				break
+			}
+		}
+
+		// Extract the sctList from the extension
```

Comment:
let's make it simpler this time, simply replace the sct extension with the value marshaled from the input `scts`, so we don't need to worry about existing extensions.